esp-hal/src/i2c.rs

```diff
@@ -1463,86 +1511,120 @@ pub trait Instance: crate::private::Sealed {
     }
 
     /// Configures the I2C peripheral for a write operation.
-    fn setup_write<'a, I>(&self, addr: u8, bytes: &[u8], cmd_iterator: &mut I) -> Result<(), Error>
+    /// - `addr` is the address of the slave device.
+    /// - `bytes` is the data two be sent.
+    /// - `start` indicates whether the operation should start by a START
+    ///   condition and sending the address.
+    /// - `cmd_iterator` is an iterator over the command registers.
+    fn setup_write<'a, I>(
+        &self,
+        addr: u8,
+        bytes: &[u8],
+        start: bool,
+        cmd_iterator: &mut I,
+    ) -> Result<(), Error>
     where
         I: Iterator<Item = &'a COMD>,
     {
-        if bytes.len() > 254 {
+        let max_len = if start { 254usize } else { 255usize };
+        if bytes.len() > max_len {
             // we could support more by adding multiple write operations
             return Err(Error::ExceedingFifo);
         }
 
+        let extra_len = if start { 1 } else { 0 };
         // WRITE command
         add_cmd(
             cmd_iterator,
             Command::Write {
                 ack_exp: Ack::Ack,
                 ack_check_en: true,
-                length: 1 + bytes.len() as u8,
+                length: extra_len + bytes.len() as u8,
             },
         )?;
 
         self.update_config();
 
-        // Load address and R/W bit into FIFO
-        write_fifo(
-            self.register_block(),
-            addr << 1 | OperationType::Write as u8,
-        );
-
+        if start {
+            // Load address and R/W bit into FIFO
+            write_fifo(
+                self.register_block(),
+                addr << 1 | OperationType::Write as u8,
+            );
+        }
         Ok(())
     }
 
     /// Configures the I2C peripheral for a read operation.
+    /// - `addr` is the address of the slave device.
+    /// - `buffer` is the buffer to store the read data.
+    /// - `start` indicates whether the operation should start by a START
+    ///   condition and sending the address.
+    /// - `will_continue` indicates whether there is another read operation
+    ///   following this one and we should not nack the last byte.
+    /// - `cmd_iterator` is an iterator over the command registers.
     fn setup_read<'a, I>(
         &self,
         addr: u8,
         buffer: &mut [u8],
+        start: bool,
+        will_continue: bool,
         cmd_iterator: &mut I,
     ) -> Result<(), Error>
     where
         I: Iterator<Item = &'a COMD>,
     {
-        if buffer.len() > 254 {
+        let (max_len, initial_len) = if will_continue {
+            (255usize, buffer.len())
+        } else {
+            (254usize, buffer.len() - 1)
```

Comment:
Can you check (would it make sense to check?) if `buffer.len() == 0`?

Reply:
Ahhh - nice catch on this edge case.

Reply:
So I made a change to deal with this but I don't think the hardware supports reading 0 bytes. From the TRM about the command register `byte_num` field:

> byte_num: Specifies the length of data (in bytes) to be read or written. Can range from 1 to 255 bytes. This bit is ignored during RSTART, STOP and END conditions.

Reply:
In testing I get timeouts from `self.wait_for_completion(!stop)`

Reply:
I was thinking we could return an error instead, the point is that I'm not really a fan of the wrapping to usize::MAX in release mode if the user decides to pass an empty buffer.

Reply:
the update changed that so there would be no wrap.

Maybe we want to detect this condition earlier like in `read_operation()` and `write_operation()` and return an error if the conditions would issue a read or a write that would not transfer any bytes:
- read_operation where buffer length is 0
- write_operation where start=false and buffer length is 0

Reply:
actually doable in setup_read/write too I think

Reply:
I think one could also argue that reading/writing (in the start=false case) 0 could also be an automatic "Ok" instead of an error - there's no need to transfer anything, after all. I'm not entirely sure, and I'm also not sure if it matters in practice.

Reply:
I'd probably prefer Ok but I would like to hear other people's input on this.

Reply:
+1 to returning `Ok` on an empty buffer

Reply:
The Ok path gets complicated when you look at a transaction:

- Write(address)
- Read(len=0)
- Read(len=10)
The ok path would skip the first read and treat the second read as a continue and not write the address as read!

Reply:
Maybe we can filter these operations where `buffer.len() == 0` out in the case of a transaction?

Reply:
that makes the look ahead to the next operation messy.  Its currently a peak on the iterator.  What if that operation gets skipped?  Or the next 4 get skipped followed by a change from read to write?

Reply:
I was thinking if we add a filter to this iterator: https://github.com/liebman/esp-hal/blob/c313989714a843d3d665a0cd4a34f6baac8ec487/esp-hal/src/i2c.rs#L295 it would be like they were never there in the first place, which should avoid the issues you mentioned above?

Reply:
Ahh [filter](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter)

Reply:
@MabezDev can we leave 0 length writes?  They are handy for probing the bus to see if a device exists.

I'm leaving the new InvalidZeroLength error in place to incase someone manages to pass in a 0 length buffers to the internal methods.  Without this error it would return a Timeout error - this way they know why.

So when someone issues a transaction like:

- Write(len=4)
- Write(len=0) # no start here and that would be an issue.

it would return an InvalidZeroLength error.

And this would still allow using embedded-hal methods to probe the bus with transactions like:

- Write(0)

And this is what the embedded-hal `write` method implements internally.

Reply:
ok - that does not quite work - no way to know when t send the stop condition :-(

Reply:
This has an issue in that with a trailing 0 length write no stop condition is sent :-(.

Reply:
Ok - I think I have this figured out - will be testing for a while

Reply:
Done.  And I've [updated my test](https://gist.github.com/liebman/4db2358c59b5147af009d699fc4cc520) code with more test cases.

- I observed each of the test cases on a logic analyzer
- similar tests should be added to the new HIL i2c tests
- one of the checks should be validating that SDA is not being held low after each transaction was run
  - that usually indicates a missing stop condition
- maybe also count the SCL transitions for each test where possible via PCNT

Reply:
I haven't checked what you decided for 0 length writes but the [lcd_cam_ov2640](https://github.com/esp-rs/esp-hal/blob/main/examples/src/bin/lcd_cam_ov2640.rs) example relies on it being available.

Reply:
I think it will be fine but I'll test. (I think I have a compatible boad somewhere).

Reply:
I looked at the usage (write of 0 bytes to probe existence of the device) and that was already included in [my set of tests](https://gist.github.com/liebman/4db2358c59b5147af009d699fc4cc520).